src/runtime/runtime.cs

```diff
@@ -133,7 +133,10 @@ internal static void Initialize(bool initSigs = false, ShutdownMode mode = Shutd
                 // If we're coming back from a domain reload or a soft shutdown,
                 // we have previously released the thread state. Restore the main
                 // thread state here.
-                PyGILState_Ensure();
+                if (mode != ShutdownMode.Extension)
```

Comment:
Instead of adding more modes, I suggest refactoring of `Initialize` into the part, that does `Py_Initialize`, and the part, that performs `Python.Runtime` initialization under assumption Python runtime has already been initialized.

Same for `Shutdown`

Reply:
I can look into this in a separate PR. For now, I think this solution is enough, but I agree that this shouldn't go into a release. The problem with such a refactoring is mainly (for me) that I'll have to refactor a lot of code that I currently understand insufficiently.

Reply:
Let's open a task then. Maybe I'll pick it up after this is in.

Reply:
Added new ticket #1378.